feat(genai): add configurable base_url for OpenAI API integration

skrashevich opened the pull request:
### PR Description: Add Configurable  base_url  for OpenAI API Integration      
                                                                                  
  #### Overview                                                                   
                                                                                  
  This pull request introduces a new feature that enhances the integration with the OpenAI API by allowing the  base_url  to be configurable. This flexibility is particularly useful for scenarios where the default OpenAI API endpoint needs to be changed or for testing against mock servers.                              
                                                                                  
  #### Changes Made                                                               
                                                                                  
  • Configuration Update: Modified the configuration file  genai.md  to include   base_url  as a new configurable parameter. The default value is set to https://api.openai.com/v1 .                                                     
  • Code Enhancements: Updated the  OpenAIClient  class in  frigate/genai/openai.py  to support the new  base_url configuration parameter:                      
    • If  base_url  is not provided or is empty, it defaults to  https://api.openai.com/v1 .                                                               
    • The  OpenAI  client is now initialized with both  api_key  and  base_url .  
                                                                                  
                                                                                  
  #### Benefits                                                                   
                                                                                  
  • Increased Flexibility: Users can now specify a different API endpoint if needed, which enhances the adaptability of the integration to different deployment environments or testing needs.                                       
  • Backward Compatibility: Existing configurations continue to work without modification, thanks to the default value setting.                              
                                                                                  
  #### Files Changed                                                              
                                                                                  
  •  docs/docs/configuration/genai.md : Added documentation for the new  base_url configuration option.                                                           
  •  frigate/genai/openai.py : Updated the OpenAI client initialization to incorporate the configurable  base_url .                                        
                                                                                  
  #### Use Case                                                                   
                                                                                  
  Users who wish to utilize an alternative endpoint or a self-hosted proxy for the OpenAI API can now simply adjust the  base_url  parameter in their configuration settings, making the client versatile for various setups.  

## Checklist

<!--
  Put an `x` in the boxes that apply.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] The code has been formatted using Ruff (`ruff format frigate`)

skrashevich committed:
feat(genai): add configurable base_url for OpenAI API integration

hawkeye217 commented on frigate/genai/openai.py:
```diff
@@ -21,7 +21,13 @@ class OpenAIClient(GenAIClient):
-        return OpenAI(api_key=self.genai_config.api_key)
+
+        if self.genai_config.base_url is None or self.genai_config.base_url == "":
+            self.genai_config.base_url = "https://api.openai.com/v1"
```
I don't think we should be managing the default url for OpenAI here, I'd prefer to let the upstream `openai-python` package specify the default if it is not set.

https://github.com/openai/openai-python/blob/52357cff50bee57ef442e94d78a0de38b4173fc2/src/openai/_client.py#L121

hawkeye217 commented on docs/docs/configuration/genai.md:
```diff
@@ -112,6 +112,7 @@ genai:
   provider: openai
   api_key: "{FRIGATE_OPENAI_API_KEY}"
   model: gpt-4o
+  base_url: https://api.openai.com/v1
```
Including the `base_url` here implies it is required. It would be better to remove this and add a line below saying that `base_url` can be provided as an optional parameter.

skrashevich committed:
docs(configuration): document optional base_url parameter for OpenAI

Add explanation for the optional base_url parameter in genai.md, allowing custom endpoint configuration for OpenAI API. Update logic in openai.py to handle empty string by setting base_url to None.

hawkeye217 reviewed (approved):
I can't test this directly, but if you've tested it with and without a `base_url` and it works, I think this can be merged. Please confirm.

isaac-mcfadyen commented:
Is there any movement on this? What needs to happen to get this merged?

I believe this is a fairly important feature because it effectively opens up *any* GenAI provider that supports the OpenAI endpoint format.

hawkeye217 commented:
In order to be merged, it needs confirmation from the contributor and/or several other users that it works with and without a `base_url` defined.

isaac-mcfadyen commented:
I just pulled and tested this and confirm it works, both with a `base_url` specified (in which case it uses the custom URL) and with no `base_url` (in which case it uses the OpenAI API).

I had to pull the main Frigate repo and apply the changes on top as it seems this PR is quite old and so the forked repo is missing some of the more recent commits in the main Frigate repo.

NickM-27 commented:
At this point this PR will likely go into 0.16

isaac-mcfadyen commented:
> At this point this PR will likely go into 0.16

🤔 is there a specific reason for that? I see that 15.0 stable isn't out quite yet and this is a fairly minor (and non-breaking) change.

NickM-27 commented:
0.15 is in RC, so besides bug fixes there likely won't be any changes. Otherwise we could go on forever with small adjustments that aren't larger features. 

We can see what the other maintainers think if they don't agree.

hawkeye217 commented:
I agree, we are in RC so this should go into 0.16.

NickM-27 commented:
So the url should be configurable by setting the `OPENAI_BASE_URL` env variable, I don't think this is actually needed

isaac-mcfadyen commented:
> So the url should be configurable by setting the `OPENAI_BASE_URL` env variable, I don't think this is actually needed

That indeed worked, thanks! Maybe this should be converted to a documentation PR then so users who want to do the same in the future don't have to look through the `python-openai` docs?

hawkeye217 commented:
Great, thanks for confirming. We'll close this PR and make a new PR for the documentation.
